### PR TITLE
pass `unused_extern_crates` in `librustdoc::doctest::make_test`

### PR DESCRIPTION
**Pull request by onur-ozkan**
blocker for https://github.com/rust-lang/rust/pull/106621

### CODE REVIEW

File: src/librustdoc/doctest.rs [hunk 1]
```diff
@@ -679,8 +679,11 @@ pub(crate) fn make_test(
             // parse the source, but only has false positives, not false
             // negatives.
             if s.contains(crate_name) {
+                // Allow `unused_extern_crates` lint rule to avoid lint errors/warnings
+                prog.push_str(&format!("#[allow(unused_extern_crates)]\n"));
+
                 prog.push_str(&format!("extern crate r#{crate_name};\n"));
-                line_offset += 1;
+                line_offset += 2;
```
Comment (ozkanonur) [inline src/librustdoc/doctest.rs:686]: I am not sure about this

**Comment by Nilstrieb** [comment]
Can you add the reasoning behind allowing this in the comment? It doesn't explain why rustdoc should do this.

**Comment by ozkanonur** [comment]
> Can you add the reasoning behind allowing this in the comment? It doesn't explain why rustdoc should do this.

This is needed for using `unused_extern_crates` on compiler docs.

ref err: https://github.com/rust-lang/rust/actions/runs/4797403514/jobs/8534349146?pr=106621#step:26:3011

will extend the comment, I guess it's not clear enough.

**Comment by Nilstrieb** [comment]
It still doesn't include reasoning why rustdoc should pass this by default. Why is not allowing it a problem?

**Comment by ozkanonur** [comment]
> It still doesn't include reasoning why rustdoc should pass this by default. Why is not allowing it a problem?

We want to add `rust_2018_idioms` lint group to the doctests.

Because of this line https://github.com/rust-lang/rust/blob/ca118d4f3e1366dee7a51a3177a2e82d9007f0a8/src/librustdoc/doctest.rs#L686

adds lines like `extern crate r#rustc_graphviz;` to the compiler tree, we are getting errors on doctests by the `rust_2018_idioms` group. Since it's automagically added, couldn't find any better way to overcome it.

**Comment by Nilstrieb** [comment]
"rustdoc implicitly inserts an `extern crate` item for the own crate which may be unused, so we need to allow the lint"
Something like that would make a good comment to add

**Pull request merged**